span.go

```diff
@@ -21,6 +24,48 @@ type Span interface {
 	// If there is a pre-existing tag set for `key`, it is overwritten.
 	SetTag(key string, value interface{}) Span
 
+	// Sets the end timestamp and calls the `Recorder`s RecordSpan()
+	// internally.
+	//
+	// Finish() should be the last call made to any span instance, and to do
+	// otherwise leads to undefined behavior.
+	Finish()
+
+	// SetTraceAttribute sets a key:value pair on this Span that also
+	// propagates to future Span children.
+	//
+	// SetTraceAttribute() enables powerful functionality given a full-stack
+	// opentracing integration (e.g., arbitrary application data from a mobile
+	// app can make it, transparently, all the way into the depths of a storage
+	// system), and with it some powerful costs: use this feature with care.
+	//
+	// IMPORTANT NOTE #1: SetTraceAttribute() will only propagate trace
+	// attributes to *future* children of the Span.
+	//
+	// IMPORTANT NOTE #2: Use this thoughtfully and with care. Every key and
+	// value is copied into every local *and remote* child of this Span, and
+	// that can add up to a lot of network and cpu overhead.
+	//
+	// IMPORTANT NOTE #3: Trace attributes keys have a restricted format:
+	// implementations may wish to use them as HTTP header keys (or key
+	// suffixes), and of course HTTP headers are case insensitive.
+	//
+	// As such, `restrictedKey` MUST match the regular expression
+	// `(?i:[a-z0-9][-a-z0-9]*)` and is case-insensitive. That is, it must
+	// start with a letter or number, and the remaining characters must be
+	// letters, numbers, or hyphens. See CanonicalizeTraceAttributeKey(). If
+	// `restrictedKey` does not meet these criteria, SetTraceAttribute()
+	// results in undefined behavior.
+	//
+	// Returns a reference to this Span for chaining, etc.
+	SetTraceAttribute(restrictedKey, value string) Span
```

Comment:
I do like like that we don't need to think about context as a user of a span.

For zipkin users to propagate debug, they'd need to access something like this directly or indirectly.

We probably need to be a bit less chatty with docs, or move them somewhere else, since this is now on a user type. For example, users should be more encouraged about baggage that isn't a trace identifier (eventhough that is supported).